pkg/addon/render.go

```diff
@@ -0,0 +1,278 @@
+		return nil, err
+	}
+	// If the name of component has been set, just keep it, otherwise will set with file name.
+	if len(comp.Name) == 0 {
+		fileName := strings.ReplaceAll(cueTemplate.Name, path.Ext(cueTemplate.Name), "")
+		comp.Name = strings.ReplaceAll(fileName, ".", "-")
+	}
+	return &comp, nil
+}
+
+// RenderApp render a K8s application
+func RenderApp(ctx context.Context, addon *InstallPackage, k8sClient client.Client, args map[string]interface{}) (*v1beta1.Application, error) {
+	if args == nil {
+		args = map[string]interface{}{}
+	}
+	app, err := generateAppFramework(addon, args)
+	if err != nil {
+		return nil, err
+	}
+	app.Spec.Components = append(app.Spec.Components, renderNeededNamespaceAsComps(addon)...)
+
+	resources, err := renderResources(addon, args)
+	if err != nil {
+		return nil, err
+	}
+	app.Spec.Components = append(app.Spec.Components, resources...)
+
+	// for legacy addons those hasn't define policy in template.cue but still want to deploy runtime cluster
+	// attach topology policy to application.
+	if checkNeedAttachTopologyPolicy(app, addon) {
+		if err := attachPolicyForLegacyAddon(ctx, app, addon, args, k8sClient); err != nil {
+			return nil, err
+		}
+	}
+	return app, nil
+}
+
+func attachPolicyForLegacyAddon(ctx context.Context, app *v1beta1.Application, addon *InstallPackage, args map[string]interface{}, k8sClient client.Client) error {
+	deployClusters, err := checkDeployClusters(ctx, k8sClient, args)
+	if err != nil {
+		return err
+	}
+	if isDeployToRuntime(addon) {
+		if len(deployClusters) == 0 {
+			// empty cluster args deploy to all clusters
+			clusterSelector := map[string]interface{}{
+				// empty labelSelector means deploy resources to all clusters
+				ClusterLabelSelector: map[string]string{},
+			}
+			properties, err := json.Marshal(clusterSelector)
+			if err != nil {
+				return err
+			}
+			policy := v1beta1.AppPolicy{
+				Name:       "deploy-addon-to-clusters",
+				Type:       v1alpha1.TopologyPolicyType,
+				Properties: &runtime.RawExtension{Raw: properties},
+			}
+			app.Spec.Policies = append(app.Spec.Policies, policy)
+		} else {
+			var found bool
+			for _, c := range deployClusters {
+				if c == multicluster.ClusterLocalName {
+					found = true
+					break
+				}
+			}
+			if !found {
+				deployClusters = append(deployClusters, multicluster.ClusterLocalName)
+			}
+			// deploy to specified clusters
+			if app.Spec.Policies == nil {
+				app.Spec.Policies = []v1beta1.AppPolicy{}
+			}
+			body, err := json.Marshal(map[string][]string{types.ClustersArg: deployClusters})
+			if err != nil {
+				return err
+			}
+			app.Spec.Policies = append(app.Spec.Policies, v1beta1.AppPolicy{
+				Name:       "specified-addon-clusters",
+				Type:       v1alpha1.TopologyPolicyType,
+				Properties: &runtime.RawExtension{Raw: body},
+			})
+			// addon should not contain workflow, this also update legacy addon with deploy2runtime steps
+			app.Spec.Workflow = nil
```

Comment:
how do the users know this omit here? Will it become a black box bug?

Reply:
This is legacy logic, I just refactor this func to a single file, but I think it's fine to remove this line.